File: sanghyunbang/1주차/Lv2/주차_요금_계산.java
// https://school.programmers.co.kr/learn/courses/30/lessons/92341

import java.util.*;

class Solution {
    
    public Map<Integer, Integer> accTime = new HashMap<>();
    public int[][] refinedRec;
    public int maxT = 23 * 60 + 59;
    
    public int[] solution(int[] fees, String[] records) {
        
        
        int baseMin = fees[0];
        int baseFee = fees[1];
        int unitMin = fees[2];
        int unitCost = fees[3];
        
        // 0-1. 여기에서 순회하면서 자동차 번호를 집어 넣기
        for (String st : records) {
            String[] arr = st.split(" ");
            accTime.put(Integer.parseInt(arr[1]), 0); // 일단은 0으로 초기화
        }
        
        // 0-2. refined 초기화 해놓기
        
        int rn = records.length;
        refinedRec = new int[rn][3];
        
        for (int i = 0; i < rn; i++) {
            String s = records[i];
            String[] arr = s.split(" ");
            
            String[] t = arr[0].split(":");
            int tt = Integer.parseInt(t[0]) * 60 + Integer.parseInt(t[1]);
            
            int carNum = Integer.parseInt(arr[1]);
            
            // 입차는 0 출차는 1
            int type = arr[2].equals("IN") ? 0 : 1;
            
            int[] input = {tt, carNum, type};
            
            refinedRec[i] = input;
        }
        
        List<Integer> carNums = new ArrayList<>(accTime.keySet());
        Collections.sort(carNums);
        int answer[] = new int[carNums.size()];
        // 이제 여기서 순회하면서 찾아보기. 다 순회해도 1000*1000이라 괜찮음
        
        for (int i = 0; i < answer.length; i++){
            int car = carNums.get(i);
            
            // 계산 넣기
            cal(car);
            
            int accT = accTime.get(car);
            int cost = baseFee + (int) Math.ceil((double) (accT - baseMin) / unitMin) * unitCost;
            
            if (accT <= baseMin) {
                answer[i] = baseFee;
            } else {
                answer[i] = cost;
            }
        }
        
        return answer;
    }
    
    // 특정 차량의 누적 시간 채우기
    public void cal(int carNum) {
        
        Deque<Integer> dq = new ArrayDeque<>();
        
        for (int[] data : refinedRec) {
            
            if (data[1] != carNum) continue;
            
            if (data[2] == 0) {
                dq.offer(data[0]);
            } else {
                accTime.put( carNum,  accTime.get(carNum) + (data[0] - dq.poll()) );
            }
        }
        
        if(!dq.isEmpty()) {
            accTime.put( carNum, accTime.get(carNum) + ( maxT - dq.poll()) );
        }
    }
}

File: seyiclover/1주차/Lv2/주차_요금_계산.py
# https://school.programmers.co.kr/learn/courses/30/lessons/92341
import math

def minute_transform(time):
    hour, minute = time.split(":")
    return int(hour) * 60 + int(minute)


def solution(fees, records):
    records_dict = {}
    answer = []

    # 요금 정보
    basic_time, basic_fee, unit_time, unit_fee = fees

    # 입출차 기록 처리
    for record in records:
        time, car_num, status = record.split()

        # 처음 등장한 차량
        if car_num not in records_dict:
            records_dict[car_num] = {
                "in_time": None,
                "total_time": 0
            }

        if status == "IN":
            records_dict[car_num]["in_time"] = minute_transform(time)

        elif status == "OUT":
            records_dict[car_num]["total_time"] += (
                minute_transform(time)
                - records_dict[car_num]["in_time"]
            )

            # 출차했으므로 현재 주차 중이 아님
            records_dict[car_num]["in_time"] = None

    # 출차 기록이 없는 차량 → 23:59 출차 처리
    end_time = minute_transform("23:59")

    for car_num, value in records_dict.items():
        if value["in_time"] is not None:
            value["total_time"] += end_time - value["in_time"]
            value["in_time"] = None

    # 차량 번호 순서대로 요금 계산
    for car_num in sorted(records_dict):
        total_time = records_dict[car_num]["total_time"]

        # 기본 시간 이하
        if total_time <= basic_time:
            fee = basic_fee

        # 기본 시간 초과
        else:
            extra_time = total_time - basic_time
            fee = basic_fee + math.ceil(extra_time / unit_time) * unit_fee

        answer.append(fee)

    return answer